Use clustermanager lib wrapped by go lib for managing cluster lifecycle

chaodaiG opened the pull request:
This PR utilizes clustermanger lib for all GKE and Boskos related logics, aiming at retiring kubetest.

This change is meant to be applied to test-infra only at the moment, let it bake for a while before enabling it in other repo(s). So instead of replacing kubetest logic in `scripts/e2e-tests.sh`, the changes are in `test/e2e-tests.sh`, by overriding `create_test_cluster` and `setup_test_cluster`. When we feel good about it we can move these 2 functions to `scripts/e2e-tests.sh` for the migration.

Update:
    `metahelper` and `prow-cluster-operation` were moved to pkg, and the only change left in this PR is `test/e2e-tests.sh` and related deps in `Gopkg.toml`

https://github.com/knative/test-infra/pull/1460

/cc @adrcunha 
/cc @chizhg

chaodaiG commented:
/hold
This PR might need to be broken down into a few smaller PRs

chizhg commented on test/prow-cluster-operation/README.md:
```diff
@@ -0,0 +1,53 @@
+### Get
+
+1. Acquiring cluster if kubeconfig already points to it
+1. If cluster name is defined then getting cluster by it's name
```
```suggestion
1. If cluster name is defined then getting cluster by its name
```

adrcunha replied:
```suggestion
1. If cluster name is defined then getting cluster by its name
```

chizhg commented on test/metahelper/main.go:
```diff
@@ -0,0 +1,42 @@
+
+	var res string
+	switch {
+	case *getKeyOpt != "" && *saveKeyOpt != "":
```
The `*`s are annoying, how about we change `flag.String` to `flag.StringVar`?

chaodaiG replied:
as discussed offline, we are not using arg[0], will explore StringVar

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+				minNodes = strconv.FormatInt(np.Autoscaling.MinNodeCount, 10)
+				maxNodes = strconv.FormatInt(np.Autoscaling.MaxNodeCount, 10)
+			} else {
+				log.Printf("DEBUG: nodepool is default-pool but autoscaling is not on: '%+v'", np)
```
I think there should be a way to get the current node count if autoscaling if off.

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+# Read metadata.json and get value for key
+# Parameters: $1 - Key for metadata
+function get_meta_value() {
+  # local metadata="${ARTIFACTS}/metadata.json"
```
Debug? If so, remove.

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+function get_meta_value() {
+  # local metadata="${ARTIFACTS}/metadata.json"
+  # cat ${metadata} | grep -oEi "\"${1}\" ?: ?\".*\"" | cut -d "\"" -f 4
+  go run knative.dev/test-infra/test/metahelper --get $@
```
Header mentions $1, here you're passing all arguments. Which one is true?

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+	var e2eRegion, e2eZone string
+	geoKey := "E2E:REGION"
+	e2eRegion = cluster.Location
+	locParts := strings.Split(e2eRegion, "_")
```
Should the separator be `-` instead of `_`? And why don't we use `RegionZoneFromLoc` in `gke` package?

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+# Run a prow-cluster-operation tool, installing it first if necessary.
+#             $1..$n - parameters passed to the tool.
+function run_prow_cluster_tool() {
+  go run knative.dev/test-infra/test/prow-cluster-operation $@
```
Either use `run_go_tool` to make the "installing it first" part true, of remove it from the comment.

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+		e2eRegion = locParts[0]
+		e2eZone = locParts[1]
+		geoKey = "E2E:ZONE"
+	} else if len(locParts) > 3 {
```
What if `len(locParts) == 3`?

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+
+# Add function call to trap
+# Parameters: $1 - Function to call
+#             $2...%n - Signals for trap
```
```suggestion
#             $2...$n - Signals for trap
```

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+#             $2...%n - Signals for trap
+function add_trap {
+  local cmd=$1
+  shift 1
```
```suggestion
  shift
```

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+function add_trap {
+  local cmd=$1
+  shift 1
+  for trap_signal in "$@"; do
```
If you quote, it will be a single entry.

```suggestion
  for trap_signal in $@; do
```

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+		"E2E:Project":  project,
+	} {
+		err = c.Set(key, val)
+		if err != nil {
```
The err can be inlined.

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+
+	for key, val := range map[string]string{
+		geoKey:         cluster.Location,
+		"E2E:Region":   e2eRegion,
```
Define these keys as constants?

adrcunha commented on test/prow-cluster-operation/README.md:
```diff
@@ -0,0 +1,53 @@
+## prow-cluster-operation
```
Since this is a temporary tool, I'm not gonna be picky about it.

adrcunha commented on test/prow-cluster-operation/actions/delete.go:
```diff
@@ -0,0 +1,46 @@
+	if err = gkeOps.Delete(); err != nil {
+		log.Fatalf("Failed deleting cluster: '%v'", err)
+	}
+	if out, err := common.StandardExec("kubectl", "config", "unset", "current-context"); err != nil {
```
So you unset the current context... What's the trick here when it fails?

chaodaiG replied:
Comments added here

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+	} {
+		err = c.Set(key, val)
+		if err != nil {
+			log.Fatalf("failed saving metadata %q:%q. err: '%v'", key, val, err)
```
As one of Go best practices, and the convention in Knative code, the log message should be capitalized, while the error message should start with a lowercase.

adrcunha replied:
```suggestion
			log.Fatalf("failed saving metadata %q:%q: '%v'", key, val, err)
```

adrcunha commented on test/prow-cluster-operation/main.go:
```diff
@@ -0,0 +1,45 @@
+	case *get:
+		actions.Get(o)
+	default:
+		log.Fatal("Must passing one of: --create, --delete, --get")
```
```suggestion
		log.Fatal("Must pass one of --create, --delete, --get")
```

chaodaiG replied:
done

adrcunha commented on test/prow-cluster-operation/main.go:
```diff
@@ -0,0 +1,45 @@
+func main() {
+	create := flag.Bool("create", false, "Create cluster")
+	delete := flag.Bool("delete", false, "Delete cluster")
+	get := flag.Bool("get", false, "Get cluster")
```
"Get cluster" from boskos ?

chaodaiG replied:
expanded the help message

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+
+	gkeClient := clustermanager.GKEClient{}
+	clusterOps := gkeClient.Setup(o.Request)
+	gkeOps := clusterOps.(*clustermanager.GKECluster)
```
I'm a bit curious why we don't directly return the `*clustermanager.GKECluster` type from `gkeClient.Setup(o.Request)` since `clusterOps` is only used in this cast.

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+	// TODO(chaodaiG): this probably should also be part of clustermanager lib
+	if out, err := common.StandardExec("gcloud", "beta", "container", "clusters", "get-credentials",
+		gkeOps.Cluster.Name, "--region", gkeOps.Cluster.Location, "--project", gkeOps.Project); err != nil {
+		log.Fatalf("failed connect to cluster: '%v', '%v'", string(out), err)
```
```suggestion
		log.Fatalf("Failed connecting to cluster: %q, '%v'", out, err)
```

adrcunha replied:
```suggestion
		log.Fatalf("failed to connect to cluster: '%v', '%v'", string(out), err)
```

chizhg commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+		log.Fatalf("failed connect to cluster: '%v', '%v'", string(out), err)
+	}
+	if out, err := common.StandardExec("gcloud", "config", "set", "project", gkeOps.Project); err != nil {
+		log.Fatalf("failed set gcloud: '%v', '%v'", string(out), err)
```
```suggestion
		log.Fatalf("Failed setting gcloud: %q, '%v'", out, err)
```

adrcunha replied:
```suggestion
		log.Fatalf("failed to set gcloud project: '%v', '%v'", string(out), err)
```

chizhg commented on test/prow-cluster-operation/actions/delete.go:
```diff
@@ -0,0 +1,46 @@
+		log.Fatalf("Failed identifying cluster for cleanup: '%v'", err)
+	}
+	log.Printf("Identified project %q and cluster %q for removal", gkeOps.Project, gkeOps.Cluster.Name)
+	// Don't wait for delete
```
It seems the comment is not correct, since the `Delete` function is a blocking operation.

chizhg commented on test/prow-cluster-operation/actions/delete.go:
```diff
@@ -0,0 +1,46 @@
+	if err = gkeOps.Delete(); err != nil {
+		log.Fatalf("Failed deleting cluster: '%v'", err)
+	}
+	if out, err := common.StandardExec("kubectl", "config", "unset", "current-context"); err != nil {
```
Can we have some comments for these two lines? It does not look intuitive to me..

chaodaiG replied:
Comments added here

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  local creation_args=""
+  (( SKIP_ISTIO_ADDON )) || creation_args+=" --addons istio"
+  [[ -n "${GCP_PROJECT}" ]] && creation_args+=" --project ${GCP_PROJECT}"
+  echo "creation args: ${creation_args}"
```
```suggestion
  echo "Creating cluster with args ${creation_args}"
```

chizhg commented on test/prow-cluster-operation/actions/get.go:
```diff
@@ -0,0 +1,27 @@
+func Get(o *options.RequestWrapper) {
+	o.Prep()
+	o.Request.SkipCreation = true
+	Create(o)
```
The logic looks strange... `SkipCreation = true` and then `Create`?

chaodaiG replied:
yes this is a bit strange, a comment is added here

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  for trap_signal in "$@"; do
+    local current_trap="$(trap -p $trap_signal | cut -d\' -f2)"
+    local new_cmd=$cmd
+    [[ -n "${current_trap}" ]] && new_cmd="${current_trap};${new_cmd}"
```
```suggestion
    [[ -n "${current_trap}" ]] && new_cmd="${current_trap};(${new_cmd})"
```

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  # Since this function is called assuming cluster creation, force removing
+  # cluster afterwards, cluster-remover by default forces cleaning up everything
+  # TODO(chaodaiG) calling async method so that it doesn't wait
+  add_trap "(run_prow_cluster_tool --delete > /dev/null &)" EXIT SIGINT
```
```suggestion
  add_trap "run_prow_cluster_tool --delete > /dev/null &" EXIT SIGINT
```

chizhg commented on test/prow-cluster-operation/main.go:
```diff
@@ -0,0 +1,45 @@
+	case *get:
+		actions.Get(o)
+	default:
+		log.Fatal("Must passing one of: --create, --delete, --get")
```
Can we add some check here to only allow one of these operations?

chaodaiG replied:
done

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  is_protected_cluster ${k8s_cluster} && \
+    abort "kubeconfig context set to ${k8s_cluster}, which is forbidden"
+
+  # Since this function is called assuming cluster creation, force removing
```
Not sure I understand this comment. This function is "create_test_cluster", why one would assume cluster creation? What's the importance of "cluster-remover forces cleaning up everything", isn't it expected? Finally, --delete doesn't wait for the operation to complete, why the TODO?

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  # cluster afterwards, cluster-remover by default forces cleaning up everything
+  # TODO(chaodaiG) calling async method so that it doesn't wait
+  add_trap "(run_prow_cluster_tool --delete > /dev/null &)" EXIT SIGINT
+}
```
Restore errexit and pipefail here.

chizhg commented on test/prow-cluster-operation/main.go:
```diff
@@ -0,0 +1,45 @@
+	delete := flag.Bool("delete", false, "Delete cluster")
+	get := flag.Bool("get", false, "Get cluster")
+	o := options.NewRequestWrapper()
+	o.AddOptions()
```
Can we add the options in the function `options.NewRequestWrapper()`, and remove this `AddOptions` functions?  It looks unnecessary and is adding some confusion.

chaodaiG replied:
done

chizhg commented on test/prow-cluster-operation/options/options.go:
```diff
@@ -0,0 +1,59 @@
+	Request          clustermanager.GKERequest
+	BackupRegionsStr string
+	AddonsStr        string
+	NoWait           bool
```
Seems this flag is not being used.

chaodaiG replied:
This is in prep for Async deletion, I will add the function

adrcunha commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+		e2eZone = locParts[1]
+		geoKey = "E2E:ZONE"
+	} else if len(locParts) > 3 {
+		log.Fatalf("location %q shouldn't contain more than 1 '_'", cluster.Location)
```
```suggestion
		log.Fatalf("location %q shouldn't contain more than one underscore", cluster.Location)
```

chizhg commented on test/prow-cluster-operation/options/options.go:
```diff
@@ -0,0 +1,59 @@
+	}
+}
+
+func (rw *RequestWrapper) Prep() {
```
There is actual a way in Go to define a slice of strings as the flag value, see the `Example 3` in https://golang.org/src/flag/example_test.go

chaodaiG replied:
just tried it, it doesn't quite work unless the `GKERequest` struct in `clustermanager` lib is declared to be of a type other than `[]string`, since the struct will need to have a `Set` function to make it work

adrcunha commented on test/prow-cluster-operation/actions/create.go:
```diff
@@ -0,0 +1,108 @@
+	// metadata so that following flow can understand the context of cluster, as
+	// well as for Prow usage later
+	// TODO(chaodaiG): this logic may need to be part of clustermanager lib as well
+	writeMetaData(gkeOps.Cluster, gkeOps.Project)
```
FYI metadata.json is used solely to provide debug information for Spyglass/Gubernator:

![image](https://user-images.githubusercontent.com/35786489/66890850-561a1f80-ef9c-11e9-9641-86602c40b350.png)

You might want to consider not complicating your clustermanager lib because of an informational piece of data.

adrcunha commented on test/metahelper/util/util.go:
```diff
@@ -0,0 +1,97 @@
+// Package util supports various needs for running tests
```
nit: missing boilerplate.

adrcunha commented on test/prow-cluster-operation/README.md:
```diff
@@ -0,0 +1,53 @@
+### Delete
+
+1. Acquiring cluster if kubeconfig already points to it
+1. If cluster name is defined then getting cluster by it's name
```
```suggestion
1. If cluster name is defined then getting cluster by its name
```

adrcunha commented on test/prow-cluster-operation/README.md:
```diff
@@ -0,0 +1,53 @@
+1. If no cluster is found from previous step then it fails
+1. Delete:
+    - [In Prow] Release Boskos project
+    - [Not in Prow] No-op
```
I thought it would delete the cluster?

adrcunha commented on test/metahelper/util/util.go:
```diff
@@ -0,0 +1,97 @@
+func (c *client) sync() error {
+	_, err := os.Stat(c.Path)
+	if os.IsNotExist(err) {
+		log.Println("write file")
```
If this is debug, remove. If not, this message should be more useful, like showing the location of the file.

adrcunha commented on test/metahelper/util/util.go:
```diff
@@ -0,0 +1,97 @@
+		log.Printf("Overriding meta %q:%q with new value %q", key, oldVal, val)
+	}
+	c.MetaData[key] = val
+	body, _ := json.Marshal(c.MetaData)
```
Why not calling sync() again?

adrcunha commented on test/metahelper/util/util.go:
```diff
@@ -0,0 +1,97 @@
+// Get gets val for key
+func (c *client) Get(key string) (string, error) {
+	if _, err := os.Stat(c.Path); err != nil && os.IsNotExist(err) {
+		return "", fmt.Errorf("file not exist: %q", c.Path)
```
```suggestion
		return "", fmt.Errorf("file %q doesn't exist", c.Path)
```

adrcunha commented on test/metahelper/main.go:
```diff
@@ -0,0 +1,42 @@
+package main
```
nit: missing boilerplate.

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  # If cluster admin role isn't set, this is a brand new cluster
+  # Setup the admin role and also KO_DOCKER_REPO if it is a GKE cluster
+  if [[ -z "$(kubectl get clusterrolebinding cluster-admin-binding 2> /dev/null)" && "${k8s_cluster}" =~ ^gke_.* ]]; then
+    # gcloud beta container clusters get-credentials ${e2e_cluster_name} --region ${e2e_cluster_region} --project ${e2e_project_name} # NC
```
Debug? Remove?

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  # Setup the admin role and also KO_DOCKER_REPO if it is a GKE cluster
+  if [[ -z "$(kubectl get clusterrolebinding cluster-admin-binding 2> /dev/null)" && "${k8s_cluster}" =~ ^gke_.* ]]; then
+    # gcloud beta container clusters get-credentials ${e2e_cluster_name} --region ${e2e_cluster_region} --project ${e2e_project_name} # NC
+    echo "${k8s_user} ${e2e_cluster_name} ${e2e_cluster_region} ${e2e_cluster_zone}"
```
More debug? Remove?

chaodaiG commented:
@adrcunha @chizhg , I think I have addressed all comments, left one TODO: using `DeleteAsync`(Not implemented yet) from clustermanager lib. If we all feel good about the direction I will start creating PRs moving `metahelper` and `prow-cluster-operation` packages over to pkg repo

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,148 @@
+  # The step above collects cluster metadata and writes to
+  # ${ARTIFACTS}/metadata.json file, use this information
+  echo "Cluster used for running tests: $(cat ${ARTIFACTS}/metadata.json)"
+  local e2e_cluster_name=$(get_meta_value "E2E:Machine")  # NA
```
You're only using them once here, having env vars with these values will just make readability worse.

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,150 @@
+
+# Get test cluster from kubeconfig, fail if it's protected
+function get_e2e_test_cluster() {
+
```
```suggestion
```

adrcunha commented on test/e2e-tests.sh:
```diff
@@ -25,13 +25,150 @@
+  run_prow_cluster_tool --create ${creation_args} || fail_test "failed creating test cluster"
+  # Should have kubeconfig set already
+  local k8s_cluster
+  k8s_cluster=$(get_e2e_test_cluster)
```
nit: merge the two lines.

chaodaiG replied:
this will hide the error from local and not fail when needed

adrcunha replied:
Oh, right. Sorry. :/

chaodaiG commented:
@chizhg , @adrcunha , this PR has been updated to use `prow-cluster-operation` and `metahelper` from pkg, this is updated in PR description. Needs another lgtm please

adrcunha commented:
/approve
/lgtm

chaodaiG commented:
/hold cancel
